crescent/hooks.py

```diff
@@ -0,0 +1,68 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Any, Generic, Literal, Protocol, Sequence, TypeVar, overload
+
+if TYPE_CHECKING:
+    from crescent.internal.app_command import AppCommandMeta
+    from crescent.internal.includable import Includable
+
+__all__: Sequence[str] = ("HookResult", "hook", "add_hooks")
+
+
+T = TypeVar("T", contravariant=True)
+U = TypeVar("U", contravariant=True)
+
+
+class Hookable(Protocol[T, U]):
+    @overload
+    def add_hooks(
+        self, hooks: Sequence[T], prepend: bool = False, *, after: Literal[False]
+    ) -> None:
+        ...
+
+    @overload
+    def add_hooks(
+        self, hooks: Sequence[U], prepend: bool = False, *, after: Literal[True]
+    ) -> None:
+        ...
+
+    def add_hooks(self, hooks: Sequence[T | U], prepend: bool = False, *, after: bool) -> None:
+        ...
+
+
+@dataclass
+class HookResult:
+    exit: bool = False
+
+
+class hook(Generic[T, U]):
+    def __init__(self, *callbacks: T, after: bool = False):
```

Comment:
Shouldn't this also be overloaded, so that it's `U` for `after=True` and `T` for `after=False`?